### PR TITLE
Add unified intent/slot filling model

### LINKED ISSUE #197: Add unified intent/slot filling model
Implement JointBERT.

### PR DESCRIPTION
**Pull request by IKostric**
Fixes #197

### CODE REVIEW

File: moviebot/nlu/annotation/joint_bert/joint_bert.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+        intent_logits, slot_logits = self(input_ids, attention_mask)
+
+        predicted_intent = intent_logits.argmax(dim=1).item()
+        predicted_slots = slot_logits.argmax(dim=2).squeeze()
+
+        return predicted_intent, predicted_slots
+
+    @classmethod
+    def from_pretrained(cls, path: str) -> None:
```
Comment (NoB0) [inline moviebot/nlu/annotation/joint_bert/joint_bert.py:79]: Nit: Wrong return type.

File: moviebot/nlu/annotation/joint_bert/joint_bert_train.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+            slot_acc,
+            on_step=False,
+            on_epoch=True,
+            prog_bar=True,
+            logger=True,
+        )
+
+        return val_loss
+
+    def configure_optimizers(self):
```
Comment (NoB0) [inline moviebot/nlu/annotation/joint_bert/joint_bert_train.py:151]: Nit: Missing return type.

File: moviebot/nlu/annotation/joint_bert/joint_bert_train.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+            json.dump(metadata, f)
+
+
+def parse_arguments():
+    """Parses the command line arguments."""
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--max_epochs", type=int, default=5)
+    parser.add_argument("--learning_rate", type=float, default=5e-5)
+    parser.add_argument("--weight_decay", type=float, default=0.0)
+    args = parser.parse_args()
```
Comment (NoB0) [inline moviebot/nlu/annotation/joint_bert/joint_bert_train.py:208]: Suggestion to add `--model_output_path` as an argument with `_MODEL_OUTPUT_PATH`.

File: moviebot/nlu/annotation/joint_bert/slot_mapping.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+from __future__ import annotations
```
Comment (NoB0) [inline moviebot/nlu/annotation/joint_bert/slot_mapping.py:1]: Nit: Missing file level docstring.

File: moviebot/core/core_types.py [hunk 5]
```diff
@@ -0,0 +1,6 @@
+from typing import TYPE_CHECKING, Dict, List, Union
```
Comment (NoB0) [inline moviebot/core/core_types.py:1]: Nit: Missing file level docstring.

**Review by NoB0 — CHANGES_REQUESTED** [review]
A couple of high level comments/question: 
* I would suggest to rename `test_nlu.py` to `test_rule_based_nlu.py`.
* Maybe a follow-up issue can be created to give the option to load different types of NLU (i.e., rule based, neural) based on the configuration

File: moviebot/nlu/annotation/joint_bert/dataset.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+            slot_annotations: A tuple of slot-value pairs in the text.
+
+        Returns:
+            A tuple of the intent, tokenized text, and labels.
+        """
+        tokens = self.tokenizer.tokenize(text)
+        labels = []
+
+        start_idx = 0
+        for slot_text, slot_label in slot_annotations:
```
Comment (vinaysetty) [inline moviebot/nlu/annotation/joint_bert/dataset.py:120]: Not sure if I follow what is going on here. A comment explaining the logic would be helpful. Maybe an example would help.

File: moviebot/nlu/annotation/joint_bert/joint_bert.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        self, input_ids: torch.Tensor, attention_mask: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Forward pass of the model.
+
+        Args:
+            input_ids: The input token IDs.
+            attention_mask: The attention mask.
+
+        Returns:
+            The intent and slot logits.
```
Comment (vinaysetty) [inline moviebot/nlu/annotation/joint_bert/joint_bert.py:47]: Nit: Tuple of intent and slot logits.

File: moviebot/nlu/annotation/joint_bert/joint_bert.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        """Predicts the intent and slot annotations for the given input.
+
+        Args:
+            input_ids: The input token IDs.
+            attention_mask (optional): The attention mask. Defaults to None.
+
+        Returns:
+            A tuple of the predicted intent and slot annotations.
+        """
+        if attention_mask is None:
```
Comment (vinaysetty) [inline moviebot/nlu/annotation/joint_bert/joint_bert.py:68]: `if not attention_mask:`

File: moviebot/nlu/annotation/joint_bert/joint_bert_train.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+        loss_intent = F.cross_entropy(
+            intent_logits.view(-1, self.intent_label_count),
+            intent_labels.view(-1),
+        )
+        loss_slot = F.cross_entropy(
+            slot_logits.view(-1, self.slot_label_count),
+            slot_labels.view(-1),
+            ignore_index=_IGNORE_INDEX,
+        )
+        val_loss = loss_intent + loss_slot
```
Comment (vinaysetty) [inline moviebot/nlu/annotation/joint_bert/joint_bert_train.py:109]: Some duplicated code with training_step

File: requirements.txt [hunk 10]
```diff
@@ -9,7 +9,12 @@ python-telegram-bot==13.15
...
 questionary
 dataclasses_json
-dialoguekit==0.0.9.dev0
\ No newline at end of file
+dialoguekit==0.0.9.dev0
+torch>=2.0.1
+torchvision>=0.15.2
+transformers>=4.31.0
+pytorch-lightning>=2.0.7
+wandb>=0.15.8
```
Comment (vinaysetty) [inline requirements.txt:20]: wikipedia missing?

**Review by vinaysetty — CHANGES_REQUESTED** [review]
Mostly minor nits but I didn't see tests for some of the new code. Is it already covered?

File: moviebot/core/core_types.py (hunk 5 above)
Comment (IKostric) [reply moviebot/core/core_types.py:1]: Done

File: moviebot/nlu/annotation/joint_bert/dataset.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+            )
+
+            # Remove slot annotations from the text
+            clean_text = re.sub(r"\[(.*?)\]\((.*?)\)", r"\1", annotated_example)
+
+            yield intent, clean_text, slot_annotations
+
+
+class JointBERTDataset(Dataset):
+    def __init__(self, path: str, max_length: int = 32):
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/dataset.py:56]: Done

File: moviebot/nlu/annotation/joint_bert/dataset.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+        self.data = load_yaml(path)
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/dataset.py:57]: Done

File: moviebot/nlu/annotation/joint_bert/dataset.py (hunk 6 above)
Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/dataset.py:120]: Good point. Added example.

File: moviebot/nlu/annotation/joint_bert/dataset.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
+import re
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/dataset.py:1]: Done

File: moviebot/nlu/annotation/joint_bert/joint_bert.py (hunk 7 above)
Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/joint_bert.py:47]: Done

File: moviebot/nlu/annotation/joint_bert/joint_bert.py (hunk 8 above)
Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/joint_bert.py:68]: RuntimeError: Boolean value of Tensor with more than one value is ambiguous.

It has to be compared to None explicitly.

File: moviebot/nlu/annotation/joint_bert/joint_bert.py (hunk 1 above)
Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/joint_bert.py:79]: Ups :)

Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/joint_bert.py:79]: Done

File: moviebot/nlu/annotation/joint_bert/slot_mapping.py (hunk 4 above)
Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/slot_mapping.py:1]: Done

File: moviebot/nlu/annotation/joint_bert/joint_bert_train.py (hunk 9 above)
Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/joint_bert_train.py:109]: Good point. Refactored.

File: moviebot/nlu/annotation/joint_bert/joint_bert_train.py (hunk 2 above)
Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/joint_bert_train.py:151]: Done

File: moviebot/nlu/annotation/joint_bert/joint_bert_train.py (hunk 3 above)
Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/joint_bert_train.py:208]: Done

File: moviebot/nlu/nlu.py [hunk 14]
```diff
@@ -196,41 +47,9 @@ def generate_dacts(
         Returns:
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/nlu.py:47]: Done

File: moviebot/nlu/nlu.py [hunk 15]
```diff
@@ -1,191 +1,42 @@
...
 
+    @abstractmethod
     def generate_dacts(
         self,
         user_utterance: UserUtterance,
         options: DialogueOptions,
-        dialogue_state: DialogueState,
-    ) -> List[DialogueAct]:
+        dialogue_state: DialogueState = None,
+    ):
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/nlu.py:37]: Done

File: moviebot/nlu/nlu.py [hunk 16]
```diff
@@ -253,9 +72,37 @@ def get_selected_option(
...
+
+        Returns:
+            Item constraint with titles of similar movies to the item in focus.
+        """
+        movie_title = item_in_focus[Slots.TITLE.value]
+        for term in ["film", "movie"]:
+            results = wikipedia.search(
+                f"I need a {term} similar to {movie_title}", 20
+            )
+            results = [r.split("(")[0].strip() for r in results]
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/nlu.py:98]: Yes, it was always the case as we dont have similarity function between movies. Results are fetched from wikipedia and then filtered for movies in our database. Note that this was moved from UserIntentsChecker as it doenst make sense there.

File: moviebot/nlu/nlu.py [hunk 17]
```diff
@@ -253,9 +72,37 @@ def get_selected_option(
...
+            for result in deepcopy(results):
+                if (
+                    result
+                    not in self.intents_checker.slot_values[Slots.TITLE.value]
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/nlu.py:102]: I'll add as temporary fix, but UserIntentsChecker should be refactored as only some components are needed in NeuralNLU.

File: moviebot/nlu/rule_based_nlu.py [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+from moviebot.core.utterance.utterance import UserUtterance
+from moviebot.dialogue_manager.dialogue_act import DialogueAct
+from moviebot.dialogue_manager.dialogue_state import DialogueState
+from moviebot.nlu.annotation.values import Values
+from moviebot.nlu.nlu import NLU
+from moviebot.nlu.user_intents_checker import UserIntentsChecker
+
+
+class RuleBasedNLU(NLU):
+    def __init__(self, config):
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/rule_based_nlu.py:22]: Done

File: moviebot/nlu/rule_based_nlu.py [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+        """
+        super().__init__(config)
+        self.intents_checker = UserIntentsChecker(config)
+
+    def _process_first_turn(
+        self, user_utterance: UserUtterance
+    ) -> List[DialogueAct]:
+        """Generates dialogue acts for the first turn of the conversation.
+
+        The system chceks if the user provided any voluntary preferences or
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/rule_based_nlu.py:44]: Done

File: moviebot/nlu/rule_based_nlu.py [hunk 20]
```diff
@@ -0,0 +1,262 @@
...
+                if user_dacts:
+                    return user_dacts
+        return []
+
+    def _process_recommendation_feedback(
+        self, user_utterance: UserUtterance
+    ) -> List[DialogueAct]:
+        """Processes recommendation feedback from the user. The function checks
+        if the user is rejecting the recommendation, inquiring about the
+        recommendation, or providing voluntary preferences.
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/rule_based_nlu.py:93]: Done

File: moviebot/nlu/rule_based_nlu.py [hunk 21]
```diff
@@ -0,0 +1,262 @@
...
+
+        Returns:
+            A list of dialogue acts.
+        """
+        feedback_intents = [
+            self.intents_checker.check_reject_intent,
+            self.intents_checker.check_inquire_intent,
+            self.intents_checker.check_reveal_voluntary_intent,
+            self._convert_deny_to_inquire,
+        ]
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/rule_based_nlu.py:106]: New Issue: #200

File: requirements.txt (hunk 10 above)
Comment (IKostric) [reply requirements.txt:20]: Line 9

File: moviebot/nlu/neural_nlu.py [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+        super().__init__(config)
+        self._model = JointBERT.from_pretrained(path)
+        self._tokenizer = BertTokenizerFast.from_pretrained("bert-base-uncased")
+
+    def generate_dacts(
+        self,
+        user_utterance: UserUtterance,
+        options: DialogueOptions = None,
+        dialogue_state: DialogueState = None,
+    ):
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/neural_nlu.py:43]: Done

File: moviebot/nlu/neural_nlu.py [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+        """Processes the utterance according to dialogue state and generate a
+        user dialogue act for Agent to understand.
+
+        Args:
+            user_utterance: UserUtterance class containing user input.
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/neural_nlu.py:48]: Done

File: moviebot/nlu/neural_nlu.py [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+            return selected_option
+
+        intent, slots = self.annotate_utterance(user_utterance)
+
+        constraints = []
+        operator = None
+        for slot in slots:
+            if slot["slot"] == "MODIFIER":
+                operator = self.get_constraint_operator(slot["value"])
+                continue
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/neural_nlu.py:69]: Yes, we dont make use of modifiers as slots, but as operators in `ItemConstraint`. This was the simplest logic I could think of, any suggestions are welcome. 
We should have a discussion about mapping modifiers as it is not straightforward.

[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/neural_nlu.py:69]: As is modifier modifies all following slots.

File: moviebot/nlu/neural_nlu.py [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+
+        return [DialogueAct(UserIntents[intent], constraints)]
+
+    def annotate_utterance(
+        self, user_utterance: UserUtterance
+    ) -> Tuple[str, list]:
+        """Annotates the utterance with intent and slot information.
+
+        Args:
+            user_utterance: UserUtterance class containing user input.
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/neural_nlu.py:86]: Done

File: moviebot/nlu/annotation/joint_bert/dataset.py [hunk 26]
```diff
@@ -0,0 +1,187 @@
...
+_TOKENIZER_PATH = "bert-base-uncased"
+
+
+def load_yaml(path: str) -> Dict[str, List[str]]:
+    """Loads the YAML file at the given path.
+
+    Args:
+        path: The path to the YAML file.
+
+    Returns:
```
Comment (NoB0) [inline moviebot/nlu/annotation/joint_bert/dataset.py:28]: Nit: Add raises section in docstring.

File: moviebot/nlu/annotation/joint_bert/dataset.py [hunk 27]
```diff
@@ -0,0 +1,187 @@
...
+        The data in the YAML file.
+    """
+    if not os.path.isfile(path):
+        raise FileNotFoundError(f"File not found: {path}")
+
+    with open(path) as f:
+        return yaml.safe_load(f)
+
+
+def parse_data(data: Dict[str, List[str]]) -> Generator[Tuple, None, None]:
```
Comment (NoB0) [inline moviebot/nlu/annotation/joint_bert/dataset.py:38]: Nit: not sure the return type is correct here.

File: moviebot/nlu/annotation/joint_bert/dataset.py [hunk 28]
```diff
@@ -0,0 +1,187 @@
...
+            yield intent, clean_text, slot_annotations
+
+
+class JointBERTDataset(Dataset):
+    def __init__(self, path: str, max_length: int = 32) -> None:
+        """Initializes the dataset.
+
+        Args:
+            path: The path to the YAML file containing the data.
+            max_length: The maximum length of the input sequence.
```
Comment (NoB0) [inline moviebot/nlu/annotation/joint_bert/dataset.py:66]:
```suggestion
            max_length: The maximum length of the input sequence. Defaults to 32.
```

File: moviebot/nlu/neural_nlu.py [hunk 29]
```diff
@@ -0,0 +1,163 @@
...
+        user_utterance: UserUtterance,
+        options: DialogueOptions = None,
+        dialogue_state: DialogueState = None,
+    ) -> List[DialogueAct]:
+        """Processes the utterance according to dialogue state and generate a
+        user dialogue act for Agent to understand.
+
+        Args:
+            user_utterance: User utterance class containing user input.
+            options: A list of options provided to the user to choose from.
```
Comment (NoB0) [inline moviebot/nlu/neural_nlu.py:49]:
```suggestion
            options: A list of options provided to the user to choose from. Defaults to None.
```

File: moviebot/nlu/nlu.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/nlu/nlu.py:106]: I see can you please create a follow-up issue for this.

File: moviebot/nlu/neural_nlu.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/nlu/neural_nlu.py:69]: Thanks for clearing this up, can you please create a follow-up issue for this.

**Review by NoB0 — APPROVED** [review]
LGTM. Just some follo-up issues to create based on answers to my previous questions.

**Review by vinaysetty — APPROVED** [review]
Great job. LGTM!

File: moviebot/nlu/annotation/joint_bert/dataset.py (hunk 27 above)
Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/dataset.py:38]: It should be correct, but I could make Tuple even more explicit. None in `Generator[Tuple, None, None]` refers to SendType and ReturnType. First is the type of parameter the generator can recieve, while the second is the return type generetor returns when done iterating.

File: moviebot/nlu/annotation/joint_bert/dataset.py (hunk 26 above)
Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/dataset.py:28]: Done

File: moviebot/nlu/annotation/joint_bert/dataset.py (hunk 28 above)
Comment (IKostric) [reply moviebot/nlu/annotation/joint_bert/dataset.py:66]: Done

File: moviebot/nlu/nlu.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/nlu/nlu.py:106]: Done

**Comment by IKostric** [comment]
> Mostly minor nits but I didn't see tests for some of the new code. Is it already covered? 

Created a new issue for tests

> I would suggest to rename test_nlu.py to test_rule_based_nlu.py.
Maybe a follow-up issue can be created to give the option to load different types of NLU (i.e., rule based, neural) based on the configuration

Done

**Pull request merged**